numpy/lib/function_base.py

```diff
@@ -1616,27 +1619,47 @@ def trim_zeros(filt, trim='fb'):
     >>> np.trim_zeros(a, 'b')
     array([0, 0, 0, ..., 0, 2, 1])
 
+    >>> b = np.array(['', '', '' ,'a', 'b', 'c', '', 'b', 'a', ''])
+    >>> np.trim_zeros(b)
+    array(['a', 'b', 'c', '', 'b', 'a'], dtype='<U1')
+
     The input data type is preserved, list/tuple in means list/tuple out.
 
     >>> np.trim_zeros([0, 1, 2, 0])
     [1, 2]
 
     """
-    first = 0
-    trim = trim.upper()
-    if 'F' in trim:
-        for i in filt:
-            if i != 0.:
-                break
-            else:
-                first = first + 1
-    last = len(filt)
-    if 'B' in trim:
-        for i in filt[::-1]:
-            if i != 0.:
-                break
-            else:
-                last = last - 1
+    try:
+        arr = np.asanyarray(filt, dtype=bool)
+
+    # str/bytes and structured arrays cannot be directly converted into bool arrays
+    except (TypeError, ValueError):
+        arr_any = np.asanyarray(filt)
+        _arr = arr_any.view(bool)
+        _arr.shape = arr_any.shape + (arr_any.dtype.itemsize,)
+        arr = _arr.any(axis=-1)
+
+    if arr.ndim != 1:
+        raise ValueError('trim_zeros requires an array of exactly one dimension')
+    elif not len(arr):
+        return filt
+
+    trim_upper = trim.upper()
+    len_arr = len(arr)
+    first = last = None
+
+    if 'F' in trim_upper:
+        first = arr.argmax()
+        # If `arr[first] is False` then so are all other elements
+        if not arr[first]:
+            return filt[len_arr:]
+
+    if 'B' in trim_upper:
+        last = len_arr - arr[::-1].argmax()
+        # If `last == len(arr) and arr[-1] is False` then all elements are False
+        if last == len_arr and not arr[-1]:
```

Comment:
```suggestion
        if not arr[-1]:
```
The only way it can be false is if it is last

Reply:
Note that `arr` is not trimmed (yet) at this point, so `arr[-1] is False` is very much a possibility without all other elements being `False`.

Reply:
Ah, sorry, What I meant is
```suggestion
        if not arr[last]:
```

Reply:
Ah yes, that's definitely much clearer; implemented in https://github.com/numpy/numpy/pull/16911/commits/6ea881c1d9c956378ffc2d6445126ac485f45d04.

Note that ``arr[last - 1]`` had to be used here as you'd otherwise get an IndexError if ``last == len(arr)``.